usecases/schema/cache.go

```diff
@@ -163,3 +170,103 @@ func (s *schemaCache) deleteClassState(name string) {
 	defer s.Unlock()
 	delete(s.ShardingState, name)
 }
+
+func (s *schemaCache) unsafeFindClassIf(pred func(*models.Class) bool) *models.Class {
+	for _, cls := range s.ObjectSchema.Classes {
+		if pred(cls) {
+			return cls
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) unsafeFindClass(className string) *models.Class {
+	for _, c := range s.ObjectSchema.Classes {
+		if c.Class == className {
+			return c
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) addClass(c *models.Class, ss *sharding.State) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.ShardingState[c.Class] = ss
+	s.ObjectSchema.Classes = append(s.ObjectSchema.Classes, c)
+}
+
+func (s *schemaCache) updateClass(u *models.Class, ss *sharding.State) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if c := s.unsafeFindClass(u.Class); c != nil {
+		*c = *u
+	} else {
+		return errClassNotFound
+	}
+	if ss != nil {
+		s.ShardingState[u.Class] = ss
+	}
+	return nil
+}
+
+func (s *schemaCache) addProperty(class string, p *models.Property) (models.Class, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	c := s.unsafeFindClass(class)
+	if c == nil {
+		return models.Class{}, errClassNotFound
+	}
+
+	// update all at once to prevent race condition with concurrent readers
+	src := c.Properties
+	dest := make([]*models.Property, len(src)+1)
+	copy(dest, src)
+	dest[len(src)] = p
+	c.Properties = dest
+	return *c, nil
+}
+
+// readOnlySchema returns a read only schema
+// Changing the schema outside this package might lead to undefined behavior.
+func (s *schemaCache) readOnlySchema() *models.Schema {
```

Comment:
Which kind of races is this solving? We'd still have raecs when accessing the classes itself

Reply:
In previous implementations, manager.GetSchema** simply returned a pointer to the internal schema. However, when a class was deleted, the schema.Classes slice could shrink. This posed a risk of readers referencing either a nil class or incorrect classes, essentially leading to a classic race condition.

In the new implementation, we've taken measures to ensure that the returned copy of the schema has its independent copy of the schema.Classes slice. So readers won't be affected when the original schema.Classes changes.

Reply:
makes sense, thanks!